gio/src/cancellable_future.rs

```diff
@@ -0,0 +1,186 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use crate::cancellable::CancellableExtManual;
+use crate::cancellable::CancelledHandlerId;
+use crate::prelude::CancellableExt;
+use crate::Cancellable;
+use crate::IOErrorEnum;
+use pin_project_lite::pin_project;
+use std::fmt::Debug;
+use std::fmt::Display;
+use std::future::Future;
+use std::pin::Pin;
+use std::task::Context;
+use std::task::Poll;
+
+// rustdoc-stripper-ignore-next
+/// Indicator that the [`CancellableFuture`] was cancelled.
+pub struct Cancelled;
+
+pin_project! {
+    // rustdoc-stripper-ignore-next
+    /// A future which can be cancelled via [`Cancellable`].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use futures::FutureExt;
+    /// # use gio::traits::CancellableExt;
+    /// # use gio::CancellableFuture;
+    /// let l = glib::MainLoop::new(None, false);
+    /// let c = gio::Cancellable::new();
+    ///
+    /// l.context().spawn_local(CancellableFuture::new(async { 42 }, c.clone()).map(|_| ()));
+    /// c.cancel();
+    ///
+    /// ```
+    pub struct CancellableFuture<F> {
+        #[pin]
+        future: F,
+
+        #[pin]
+        waker_handler_cb: Option<CancelledHandlerId>,
+
+        cancellable: Cancellable,
+    }
+}
+
+impl<F> CancellableFuture<F> {
+    // rustdoc-stripper-ignore-next
+    /// Creates a new `CancellableFuture` using a [`Cancellable`].
+    ///
+    /// When [`cancel`](CancellableExt::cancel) is called, the future will complete
+    /// immediately without making any further progress. In such a case, an error
+    /// will be returned by this future (i.e., [`Cancelled`]).
+    pub fn new(future: F, cancellable: Cancellable) -> Self {
+        Self {
+            future,
+            waker_handler_cb: None,
+            cancellable,
+        }
+    }
+
+    // rustdoc-stripper-ignore-next
+    /// Checks whether the future has been cancelled.
+    ///
+    /// Note that all this method indicates is whether [`cancel`](CancellableExt::cancel)
+    /// was called. This means that it will return true even if:
+    ///   * `cancel` was called after the future had completed.
+    ///   * `cancel` was called while the future was being polled.
+    #[inline]
+    pub fn is_cancelled(&self) -> bool {
+        self.cancellable.is_cancelled()
+    }
```

Comment:
Should there maybe be a getter for the `&Cancellable`?

Reply:
I agree having an accessor to the inner cancellable might be useful as generic API. 

However, I would keep the function `is_cancelled()` because I think it gives a handy shortcut and it allows the possibility to add a minimal documentation about the usage with the polling (see the current documentation).

Don't you think?

Reply:
Yes, I meant in addition to that function :)